modules/FiniteSetsExtTheorems.tla

```diff
@@ -0,0 +1,181 @@
+
+THEOREM FoldSetEmpty ==
+    ASSUME NEW op(_,_), NEW base
+    PROVE  FoldSet(op, base, {}) = base
+
+THEOREM FoldSetNonempty ==
+    ASSUME NEW op(_,_), NEW base, NEW S, S # {}, IsFiniteSet(S)
+    PROVE  \E x \in S : FoldSet(op, base, S) = op(x, FoldSet(op, base, S \ {x}))
+
+THEOREM FoldSetType ==
+    ASSUME NEW Typ, NEW op(_,_), NEW base \in Typ,
+           NEW S \in SUBSET Typ, IsFiniteSet(S),
+           \A t,u \in Typ : op(t,u) \in Typ
+    PROVE  FoldSet(op, base, S) \in Typ
+
+\* more useful than FoldSetNonempty when op is associative and commutative
+THEOREM FoldSetAC ==
+    ASSUME NEW Typ, NEW op(_,_), NEW base \in Typ, 
+           \A t,u \in Typ : op(t,u) \in Typ,
+           \A t,u \in Typ : op(t,u) = op(u,t),
+           \A t,u,v \in Typ : op(t, op(u,v)) = op(op(t,u),v),
+           NEW S \in SUBSET Typ, IsFiniteSet(S),
+           NEW x \in S
+    PROVE  FoldSet(op, base, S) = op(x, FoldSet(op, base, S \ {x}))
+
+THEOREM FoldSetACAddElement ==
+    ASSUME NEW Typ, NEW op(_,_), NEW base \in Typ, 
+           \A t,u \in Typ : op(t,u) \in Typ,
+           \A t,u \in Typ : op(t,u) = op(u,t),
+           \A t,u,v \in Typ : op(t, op(u,v)) = op(op(t,u),v),
+           NEW S \in SUBSET Typ, IsFiniteSet(S),
+           NEW x \in Typ \ S
+    PROVE  FoldSet(op, base, S \union {x}) = op(x, FoldSet(op, base, S))
+
+THEOREM FoldSetDisjointUnion ==
+    ASSUME NEW Typ, NEW op(_,_), NEW base \in Typ, 
+           \A t,u \in Typ : op(t,u) \in Typ,
+           \A t,u \in Typ : op(t,u) = op(u,t),
+           \A t,u,v \in Typ : op(t, op(u,v)) = op(op(t,u),v),
+           \A t \in Typ : op(base, t) = t,
+           NEW S \in SUBSET Typ, IsFiniteSet(S),
+           NEW T \in SUBSET Typ, IsFiniteSet(T), S \cap T = {}
+    PROVE  FoldSet(op, base, S \union T) = 
+           op(FoldSet(op, base, S), FoldSet(op, base, T))
+
+---------------------------------------------------------------------------
+
+THEOREM SumSetNat ==
+    ASSUME NEW S \in SUBSET Nat, IsFiniteSet(S)
+    PROVE  SumSet(S) \in Nat
+
+THEOREM SumSetInt ==
+    ASSUME NEW S \in SUBSET Int, IsFiniteSet(S)
+    PROVE  SumSet(S) \in Int
+
+THEOREM SumSetEmpty == SumSet({}) = 0
+
+THEOREM SumSetNonempty ==
+    ASSUME NEW S \in SUBSET Int, IsFiniteSet(S), NEW x \in S 
+    PROVE  SumSet(S) = x + SumSet(S \ {x})
+
+THEOREM SumSetAddElement ==
+    ASSUME NEW S \in SUBSET Int, IsFiniteSet(S), NEW x \in Int \ S
+    PROVE  SumSet(S \union {x}) = x + SumSet(S)
+
+THEOREM SumSetDisjointUnion ==
+    ASSUME NEW S \in SUBSET Int, IsFiniteSet(S),
+           NEW T \in SUBSET Int, IsFiniteSet(T), S \cap T = {}
+    PROVE  SumSet(S \union T) = SumSet(S) + SumSet(T)
+
+THEOREM SumSetNatSubset ==
+    ASSUME NEW S \in SUBSET Nat, IsFiniteSet(S),
+           NEW T \in SUBSET S
+    PROVE  SumSet(T) <= SumSet(S)
+
+THEOREM SumSetNatZero ==
+    ASSUME NEW S \in SUBSET Nat, IsFiniteSet(S)
+    PROVE  SumSet(S) = 0 <=> S \subseteq {0}
```

Comment:
`S \subseteq {0}` is elegant, but perhaps the more explicit `(S = {} \/ S = {0})` instead of `S \subseteq {0}` (same below for ProductSetNatOne) is easier to understand?